[11.0][FIX] Sing tax l10n es aeat sii

RodrigoBM opened the pull request:
Buenas, 

He creado la siguiente PR para solucionar un problema de signos que puede darse en según que facturas al ser notificads al SII.
Actualmente sin importar el signo de cada linea de factura, este dato pasa por una transformación en absoluto perdiendo su signo, y asignándole un signo según el tipo de factura, esto puede ser incorrecto cuando se nos presentan facturas con lineas positivas y negativas con distintas cuotas de IVA, como se muestra en el siguiente ejemplo de una factura rectificativa de proveedor:

![Selección_104](https://user-images.githubusercontent.com/6359121/94255222-328b8300-ff28-11ea-9ab9-46a886a89096.png)

Si comprobamos como ha llegado la información al SII vemos que el importe de las "CuotasSoportadas" pasan a negativo y dando un Total de "CuotaDeducible" de -3.96 y un importe total de -22.06, lo cual no concuerda con el detalle de las líneas.

![Selección_107](https://user-images.githubusercontent.com/6359121/94256771-92832900-ff2a-11ea-8b46-de433609f696.png)

Cuando lo correcto seria mantener el signo de cada una de las lineas para que tanto el desglose de las cuotas y los valores totalizados concuerden.

"TipoImpositivo": "21.0",
"BaseImponible": 6.11,
"CuotaSoportada": 1.28

"TipoImpositivo": "10.0",
"BaseImponible": -26.77,
"CuotaSoportada": -2.68

"CuotaDeducible": -1.4 (-2.68+1.28)

Un saludo

CC: @soniaviciana

RodrigoBM committed:
Se corrige error de signo al informa al sii, ademas de cambiar la asginacion de los mapeos de los impuestos compuestos para que sean los de cantidad positiva los que figuren en el mapeo.

Esta modificacion es necesaria para mantener el signo correcto de cada impuesto aplicado y dejar de aplicar el absoluto sobre este y perder la informacion del signo, arreglando
el problema de si tenemos dos lineas de iva de diferentes impuestos con importe negativo y otro positivo, en el detalle de iva actualemente se informa los dos en negativo, cuando lo correcto es informa cada uno con su signo correspondiente. Ademas el calculo sobre la cuota deducible se hacia de forma incorrecta si se parte de una cuota soportada de cada impuesto incorrecto.

pedrobaeza commented:
Hola @RodrigoBM necesito que valides #1470 y lo llevamos a 11.0 para que sobre ello construyas este parche, ya que cambia totalmente según eso.

RodrigoBM commented:
Hola @pedrobaeza voy a ver el PR que comentas.

RodrigoBM commented:
Hola @pedrobaeza ya he revisado el cambio de la PR que me has puesto y me parece correcto. Un saludo

pedrobaeza commented:
Por qué no haces entonces review y le das a aprobar para dejar constancia?

pedrobaeza commented:
Perdón, no la había visto :smiley:

pedrobaeza commented:
Ya está hecho el cherry-pick. Haz por favor rebase en esta rama y resuelve los conflictos para ajustar lo que comentas del signo y me avisas cuando esté para revisarlo con urgencia para pasar todo a v13.

RodrigoBM commented:
Buenas @pedrobaeza ya esta listo el código, hemos vuelto a realizar pruebas con resultado satisfactorio.

RodrigoBM commented:
Buenas @pedrobaeza ya esta disponible para revisarlo, gracias de antemano

pedrobaeza commented:
Hola, varias cosas:

1. Mejor todo en un solo commit.
2. El `abs` se puso porque para impuestos con hijos, uno positivo y otro negativo (por ejemplo, los intra-comunitarios), no estaba asegurado el signo que iba a tener la operación, y eso que yo sepa no ha cambiado, por lo que este cambio producirá error.

RodrigoBM commented:
Buenas @pedrobaeza vamos a ver lo que comentas en el punto dos.

SoniaViciana commented:
Buenos días @pedrobaeza 

Por favor, ¿nos puedes dar más detalles sobre el punto 2 para revisar si hay algo que no hemos tenido en cuenta?

Hemos realizado pruebas con este FIX aplicado sobre facturas de adquisiciones intracomunitarias y el resultado del envío al SII es correcto, si revisamos el desglose:

**Factura de compra:**

![image](https://user-images.githubusercontent.com/46522913/94671890-c1c3dc80-0314-11eb-97b9-00c241521bb4.png)


En el SII:

![image](https://user-images.githubusercontent.com/46522913/94671094-b3c18c00-0313-11eb-9eaa-642032acd393.png)

**Rectificativas de compra:**

![image](https://user-images.githubusercontent.com/46522913/94671589-5da11880-0314-11eb-893c-37a078ed0644.png)

En el SII:

![image](https://user-images.githubusercontent.com/46522913/94671779-96d98880-0314-11eb-84bf-960fd2f5c810.png)


Gracias por vuestra ayuda.

Saludos,

pedrobaeza commented:
No lo tengo en la memoria fresco, pero los mapeados de no todos esos impuestos compuestos devolvían el adecuado en su sitio (si tienes SP_IN_1 y SP_IN_2, unas veces uno era el positivo y otras el negativo), y eso provocaba que salieran al final con el signo equivocado. He dicho intracomunitario porque me sonaba ése pero puede ser de otro de los compuestos. Echad un vistazo a los compuestos y si nos aseguramos que todos están bien mapeados, entonces perfecto.

SoniaViciana commented:
Buenas tardes @pedrobaeza 

Hemos estado realizando pruebas y los signos de los impuestos aparecen correctos independientemente de que sean compuestos. En todo caso, el cambio sólo afectaría a cómo se desglosan los impuestos en el detalle técnico del SII, pero no en cualquier otro modelo, por lo que no debería debería surgir ningún error inesperado.

Saludos,

pedrobaeza commented:
@misern2 vosotros tuvisteis una necesidad parecida. ¿Puedes confirmar por tu parte que la resuelve?

SoniaViciana commented:
Buenos días @RodrigoBM 

Por favor, cuando puedas revisamos en más profundidad la mejora para que aplique bien a impuestos como "Adquisición de servicios extracomunitarios". En este tipo de impuestos compuestos, la aplicación del signo no se hace de forma correcta en el 100% de los casos:

![image](https://user-images.githubusercontent.com/46522913/98660132-3e4ecf80-2345-11eb-95cd-8c4bba115e57.png)

Es necesario que tanto la base imponible como las cuotas contengan siempre el mismo valor (positivo o negativo), pero que en ningún caso se mezclen.

Quién debería marcar el signo en el desglose del SII es la base imponible a nivel de línea de factura, teniendo en cuenta que en el envío técnico al SII se totalizan:

![image](https://user-images.githubusercontent.com/46522913/98660984-37748c80-2346-11eb-9c42-ad76e81d23bf.png)

Muchas gracias!

@danielduqma

pedrobaeza commented:
A eso era lo que yo me refería desde el principio, Sonia...

Para ese impuesto, habría que cambiar el mapeado del impuesto hijo (1) al (2), y revisar si hay alguno más en el mismo caso. A partir de v13, ya no existe este problema, porque he cambiado el modo en el que se calcula (debido a los cambios en la estructura de Odoo).

SoniaViciana commented:
> A eso era lo que yo me refería desde el principio, Sonia...
> 
> Para ese impuesto, habría que cambiar el mapeado del impuesto hijo (1) al (2), y revisar si hay alguno más en el mismo caso. A partir de v13, ya no existe este problema, porque he cambiado el modo en el que se calcula (debido a los cambios en la estructura de Odoo).

Muchas gracias por la observación @pedrobaeza ! lo revisamos y analizamos al detalle en todos aquellos impuestos en los que pueda darse este fallo.

Saludos!

RodrigoBM commented:
Buenas @pedrobaeza,

Ya hemos realizado los cambios de los mapeos para asignar aquellos que son de tipo compuesto correctamente.

Ademas hemos incluido los negativos de los impuestos compuestos en la fila de mapeo de NotIncludedInTotal para que el importe total a informar en el SII sea el correcto, resolviendo parte del problema detallado en la tarea https://github.com/OCA/l10n-spain/issues/1490

Esto ya lo tenemos puesto en un cliente desde hace un mes y no nos ha surgido errores. 

Un saludo

JuanjoA commented:
Hola, me acabo de encontrar este problema, no había visto el commit y me sale la misma solución que a Rodrigo, ¿cómo va el posible merge?

pedrobaeza commented on l10n_es_aeat_sii/data/aeat_sii_map_data.xml:
```diff
@@ -177,35 +177,50 @@
-              ref('l10n_es.account_tax_template_p_irpf21t'),
-              ref('l10n_es.account_tax_template_p_irpf21td'),
-              ref('l10n_es.account_tax_template_p_irpf21te'),
+                ref('l10n_es.account_tax_template_s_irpf1'),
```
No cambies esta indentación aunque no sea estrictamente correcta para no aumentar tanto el diff.

RodrigoBM replied:
Buenas @pedrobaeza ya he modificado esto para no incluir la indentación.

pedrobaeza commented on l10n_es_aeat_sii/data/aeat_sii_map_data.xml:
```diff
@@ -177,35 +177,50 @@
+                ref('l10n_es.account_tax_template_p_iva21_isp_2'),
+                ref('l10n_es.account_tax_template_p_iva10_isp_2'),
+                ref('l10n_es.account_tax_template_p_iva4_isp_2'),
+                ref('l10n_es.account_tax_template_p_iva21_sp_ex_1'),
```
Esto que añades no tiene que ver con el tema tratado (supongo que será lo otro del 01/01/2021).

RodrigoBM replied:
Si esto es para corregir el tema de las nuevas restricciones del SII que entran en vigor el 01/01/2021 comentado en https://github.com/OCA/l10n-spain/issues/1490

pedrobaeza replied:
Ya, pero no debe incluirse a la vez. Al menos en un commit aparte, y por favor pon mensajes de commit lo suficientemente explicativos.

RodrigoBM replied:
Venga, lo cambio a dos commits diferenciados y añado mas info en el commit.

RodrigoBM replied:
@pedrobaeza Cambios realizados

RodrigoBM committed:
Se añaden el mapeo de los impuestos compuestos negativos en el apartado NotIncludedInTotal para poder tener en cuenta esta cantidad en el sumatorio del importe total
y que realicen correctamente los calculos para las nuevas restricciones que aplican al 01/01/2021

pedrobaeza commented:
Para la próxima, te faltaría poner `[FIX] l10n_es_aeat_sii: <description>` en el commit para seguir las guidelines al respecto, y luego sería preferible poner en inglés dicho mensaje, pero bueno, para ésta ya no me pongo más quisquilloso:

/ocabot merge minor

RodrigoBM commented:
Perfecto @pedrobaeza, la próxima vez me ajustare a las guidelines.

Muchas gracias y un saludo
